swap/src/bitcoin/wallet.rs

```diff
@@ -7,55 +7,100 @@ use crate::{
     execution_params::ExecutionParams,
 };
 use ::bitcoin::{util::psbt::PartiallySignedTransaction, Txid};
-use anyhow::{Context, Result};
+use anyhow::{anyhow, bail, Result};
 use async_trait::async_trait;
 use backoff::{backoff::Constant as ConstantBackoff, tokio::retry};
-use bitcoin_harness::{bitcoind_rpc::PsbtBase64, BitcoindRpcApi};
-use reqwest::Url;
-use std::time::Duration;
-use tokio::time::interval;
+use bdk::{
+    blockchain::{noop_progress, Blockchain, ElectrumBlockchain},
+    electrum_client::{Client, ElectrumApi},
+    keys::GeneratableDefaultOptions,
+    FeeRate,
+};
+use reqwest::{Method, Url};
+use serde::{Deserialize, Serialize};
+use std::{path::Path, sync::Arc, time::Duration};
+use tokio::{sync::Mutex, time::interval};
+
+const SLED_TREE_NAME: &str = "default_tree";
 
-#[derive(Debug)]
 pub struct Wallet {
-    pub inner: bitcoin_harness::Wallet,
+    pub inner: Arc<Mutex<bdk::Wallet<ElectrumBlockchain, bdk::sled::Tree>>>,
     pub network: bitcoin::Network,
+    pub http_url: Url,
+    pub rpc_url: Url,
 }
 
 impl Wallet {
-    pub async fn new(name: &str, url: Url, network: bitcoin::Network) -> Result<Self> {
-        let wallet = bitcoin_harness::Wallet::new(name, url).await?;
+    pub async fn new(
+        electrum_rpc_url: Url,
+        electrum_http_url: Url,
+        network: bitcoin::Network,
+        datadir: &Path,
+    ) -> Result<Self> {
+        // todo: Implement conversion to anyhow::error so we can use ?
+        let client =
+            Client::new(electrum_rpc_url.as_str()).expect("Failed to init electrum rpc client");
```

Comment:
```suggestion
        let client =
            Client::new(electrum_rpc_url.as_str()).map_err(|e| anyhow!("Failed to init electrum rpc client: {:?}", e))?;
```
In the mean time you do not need to use `expect`.